### ISSUE TITLE
Site to site support

### ISSUE THREAD

**Issue by relaxibus** [ngoduykhanh/wireguard-ui#367]
Hi, are site to site configuration supported? I need to connect various router/subnets to the server.

**Comment by jameswillhoite** [comment]
Site to site can work with wireguard. I use it now between a Firewalla Firewall (home) and the Wireguard Server (work). In this instance, I have the Wireguard Server (work) automatically connect to my Firewalla (home). My Firewalla already takes care of the static routing to route traffic destined to my work network to the the Wireguard Server. But I had to add a Static Route to my work Firewall/Router to point traffic for my home network to route traffic to the wireguard server. 

In the Wireguard config file, I used the Allowed IPs section to specify subnet 192.168.2.0/24 to be allowed (home subnet) and my firewalla has the Allowed IPs section specify  subnet 192.168.1.0/24 (work subnet).

You have to have two different subnets to work.

**Comment by relaxibus** [comment]
Thanks! But can this be configured vie your UI?

**Comment by jameswillhoite** [comment]
Not exactly. When creating the Client Config, in the Allowed IPs section. Enter the desired subnet that is allowed to transverse the tunnel. 

In may case I created a client on my Firewalla Wireguard server. Entered in the `AllowedIPs=192.168.2.0/24` and installed it (the WireGuard config file) on my Work Wireguard Server. If you put `AllowedIPs=0.0.0.0/0` then it will tunnel ALL traffic through this tunnel, you don't want that, just the subnet you want to allow. Bring up the connection, then I had to place a static route in my Work's router that said any traffic that is destined for subnet 192.168.2.0/24 to route that traffic to ip address 192.168.1.25 (the server/computer running the Wireguard Server at work).

Since my Home Wireguard Server (Firewalla) is built into my Router I did not need to do anything there, it is already configured to do so. Otherwise, you would have to do the same thing, in your home router you need to specify a static route for your work network to point to the server/computer at home that is hosting the wireguard server, it will then route the traffic to the desired tunnel.

Home this helps!

**Comment by relaxibus** [comment]
Ah ok, sorry ;-)
Many thanks for your valuable help. 
It would be nice to have a complete Wireguard Server UI to deploy all this without the need to write a single line, for Linux dumb persons like me.

**Comment by igoooor** [comment]
Same for me, that would be great if that could be managed from the UI

**Comment by Dirk23** [comment]
+1 für Side to Side Feature!

**Comment by Tarow** [comment]
I would also highly appreciate this.
Apart from the subnet check on the `IP Allocation` input field, wouldn't this already work?

So for example:
My Wireguard server managed by Wireguard-UI uses the subnet `10.0.1.0/24`. Every client in my own home network has some IP `10.0.1.x/32`.

Now i would like to connect my friends home network, who uses the subnet `10.0.2.0/24` on his server.
Naturally i would configure the peer like this: 
![image](https://github.com/ngoduykhanh/wireguard-ui/assets/15030658/eb1b0200-0b85-4142-a3de-ea4155a790c2)

Now currently this fails with `IP 10.0.2.0 does not belong to any network addresses of WireGuard server`. If that "sanity" check was removed and the IP Allocation would just be accepted, the generated config should be fine in my opinion. Only option that would be nice when adding a site-to-site client would be an optional endpoint probably.

Am i missing something or is that actually all that would have to be done? I might start working on a PR then.

**Comment by jameswillhoite** [comment]
No. The Allocation field is the IP address that the client would use to connect to the wireguard-ui server. The allowed section would be what IP Addresses are allowed through the tunnel.

Example I have a Split-tunnel for my parents house. I would create a client in the wireguard-ui and let it pick an Allocation IP address for me (or specify one, but it should end with /32). Then in the allowed list I would put the ip addresses or subnet that is allowed to go through that tunnel.

Client 1:
Allocation=10.0.1.115/32
Allowed=192.168.1.0/24

When you install this config at your friends house, it would assign that client the ip address 10.0.1.115 and would only tunnel the traffic destined for 192.168.1.0 subnet through the tunnel and all other traffic would go through the friends own ISP.

You have to make sure no networks conflict. 

Friends Network = 192.168.1.0/24
Own Network = 192.168.2.0/24
Wireguard = 10.0.2.0/24

If you want other WG Cleints to be allowed to go through the tunnel you would have to specify those too, or put `Allowed=192.168.1.0/24,10.0.2.0/24`

**Comment by Tarow** [comment]
Thanks for the explaination, but i don't seem to understand the issue with the allocated IP yet.
So my example is a bit different.

Say i have a Wireguard Server running using the `10.0.1.0/24` subnet. I have multiple clients `10.0.1.0/32` connected to my server.
Now my friend has a similar setup, just that his Server is using the `10.0.2.0/24` Wireguard subnet, and all his clients (phone, tablet, ...) are having a `10.0.2.x/32` address.

Now i only want to connect both our sites. His server with my server. They act as bounce servers with IP forwarding enabled.
This would allow me to reach his home network (eg. `192.168.0.0/24`), but also his Wireguard clients.

So the way i imagine it is like this:
![image](https://github.com/ngoduykhanh/wireguard-ui/assets/15030658/d3c71473-fe9d-4ab5-9046-858c637faba9)

My Home Network: `10.1.1.0/24`
My Wireguard Network: `10.0.1.0/24`
Friends Home Network: `192.168.0.0/24`
Friends Wireguard Network: `10.0.2.0/24`

---

Now if i would use Wireguard-UI to generate a config for my friends server and i try to model the above image, i would put these information:
![image](https://github.com/ngoduykhanh/wireguard-ui/assets/15030658/0df4679d-d7ab-47d4-8894-31452ba28ddd)

Now, if the sanity check on the IP Allocation input field, would not be present, there would be two config files.
My servers wg0.conf would look like this:
```
[Interface]
Address = 10.0.1.0/24
ListenPort = 51820
PrivateKey = ...


# My friends server -> Requests to his WG subnet and home subnet will be routed to him
[Peer]
PublicKey = ...
PresharedKey = ...
AllowedIPs = 10.0.2.0/24,192.168.0.0/24
```

The generated config for my friends server would look like this:
```
[Interface]
Address = 10.0.2.0/24
PrivateKey = ...

# My server, only requests to my Wireguard or home subnet will be routed to me
[Peer]
PublicKey = ...
PresharedKey = ...
AllowedIPs = 10.0.1.0/24,10.1.1.0/24
Endpoint = vpn.mydomain.com:51820
PersistentKeepalive = 30
```

Now i would have thought that this configuration would be perfectly fine, am i missing something?

**Comment by jameswillhoite** [comment]
That looks good except for the Allocation Part. Think of the Client Profile as a Computer being turned on. It gets one IP address from the DHCP server ... it doesn't get the whole subnet. This is a static IP address allocation. Give it a IP address within the subnet of YOUR WireGuard Server address pool.

Is your WG Server also your Router? IF not, then you will need to put a Static Route in your router for your friends subnet to point to your WG Server... Otherwise your router will not know what to do with the packet.

**Comment by jameswillhoite** [comment]
Let me clarify a little more. The Allocate IP address is a IP address on YOUR WG Server. When you save that Config File, you download it and send it to your friend. They install the config file on their WG Server under (/etc/wireguard  if it is linux) then they can call wg-quick up client_profile and it will open the connection. This will not persist across reboot. You would have to call `systemctl enable wg-quick@client_profile.service` and then `systemctl daemon-reload` then `systemctl start wg-quick@client_profile` to start it and it will start the connection automatically. You can call `systemctl status wg-quick@client_profile` to see the current status. Fill in `client_profile` with whatever filename it is saved as.

Then Your Firend would have to do the same. Create a Profile for you, and you would do the same on your system.

**Comment by Tarow** [comment]
@jameswillhoite 

Thanks again, but what makes you think that a peer has to be in the subnet of my "server". In general, wireguard doesn't have a client-server concept right? It's pure peer to peer connections on point to point interfaces.

So since my friends already has a wireguard interface running with an IP `10.0.2.0`, i want to add exactly this peer to my 'server'. I don't want to create a second profile or have him create a new interface etc. And i don't think it is necessary either.

To give you an example, i tested a connection from a 'client' to my 'server', both in completely different subnets.
My server config:
```
[Interface]
Address = 10.0.1.0/24
ListenPort = 51820
PrivateKey = ...

[Peer]
PublicKey = ...
PresharedKey = ...
AllowedIPs = 10.5.5.1/32
```

My client config (in this case, it's my phone):
```
[Interface]
Address = 10.5.5.1/32
PrivateKey = ...

[Peer]
PublicKey = ...
PresharedKey = ...
AllowedIPs = 10.0.1.0/24, 10.1.1.0/24
Endpoint = vpn.mydomain.com:51820
PersistentKeepalive = 30
```

Now, this setup works perfectly fine. If i request a website hosted on `10.1.1.99` for example, it can reach it from my phone. I can also reach my server under `10.0.1.0` without any problems. My server can also ping the phone without any issues, despite being on a completely different subnet.

So i think my configuration from above would work perfectly fine too.

---

From what i understood, the suffix in the interface definition only affects the routes being created.
Example config:
```
[Interface]
Address = 10.0.1.0/24
ListenPort = 51820
PrivateKey = ...

[Peer]
PublicKey = ...
PresharedKey = ...
AllowedIPs = 10.0.1.1/32
```

So this means, that the computer on the Wireguard interface has the IP address `10.0.1.0`. It also means that the Wireguard interface should handle all IP addresses in the /24 subnet. So it only affects the routes, not the IP allocation. This peer has the IP `10.0.1.0`, regardless if i use `/8`, `/16`, `/24` or `/32` subnet mask in the interface definition.

In this case i configured the only peer to be in the same subnet. If i look at the routes created by wireguard i can see this:
![image](https://github.com/ngoduykhanh/wireguard-ui/assets/15030658/a129db92-60c5-4305-bac8-b45225adcc27)

There is only one route for our wg0 interface that handles the whole subnet, since all peers are part of the subnet.

Now let's define the interface as a `/32` address:
```
[Interface]
Address = 10.0.1.0/32
...
```
Checking the routes shows me, that now there is an individual route for my peer, since the interface subnet does not include my peer.
![image](https://github.com/ngoduykhanh/wireguard-ui/assets/15030658/4a2c4fbd-0dc6-4a82-93c3-a25a9fcbc870)


Now if i add another peer `192.168.2.0/24` in a completely different subnet and check the routes:
![image](https://github.com/ngoduykhanh/wireguard-ui/assets/15030658/9b176f89-fcec-4655-88b0-54aad9141f11)

--- 

So yea, i think you get my point.
I think this just shows, that you can connect any peer that you like. Regardless of the subnet that the client is using.
That's why i think the "sanity check" in the wireguard-ui, that checks if a client does belong to the same subnet of the "server" is kind of pointless, because it's an unecessary limitation with focus on a "server" that manages multiple "clients", even though wireguard is no client-server protocol.

Another example of this is the `Server Interface Addresses` setting. If you give you server an `/32` IP, eg. `10.0.1.0/32`, this is a perfectly valid setting. The only downside this has, is that wireguard will create a new route for every peer. But with this setting, wireguard-ui won't let you create any new client, because there is no other IP address in the servers subnet.
Maybe i am missing something, but i think this wireguard-ui limitation, that clients have to be in the same subnet as the server is strange and not necessary.

**Comment by jameswillhoite** [comment]
It is still a subnet. You have routing in place that allows you to ping the different devices. You can't just say, hey allow my friends network to link to my own. You have to give it a single IP address (so it knows where to route the packets) to connect to your WireGuard Server and then route his networks to the correct  interface. You are correct in that it really isn't a client to server .... you can turn on the tunnel and it will be on, but there is no connection that is routable. This is why you need to create a profile for your friend to use that will have everything they need to connect to your network, and they have to do the same for you.

**Comment by Tarow** [comment]
Yes, so going back to my comment here, i don't see why this shouldn't work.

My server config:
```
[Interface]
Address = 10.0.1.0/24
ListenPort = 51820
PrivateKey = ...


# My friends server -> Requests to his WG subnet and home subnet will be routed to him
[Peer]
PublicKey = ...
PresharedKey = ...
AllowedIPs = 10.0.2.0/24,192.168.0.0/24
Endpoint = vpn.my-friends-domain.com:51820
```

The server config for my friend, generated by wireguard-ui (assuming, it didn't have the limitations).
```
[Interface]
Address = 10.0.2.0/24
PrivateKey = ...

# My server, only requests to my Wireguard or home subnet will be routed to me
[Peer]
PublicKey = ...
PresharedKey = ...
AllowedIPs = 10.0.1.0/24,10.1.1.0/24
Endpoint = vpn.mydomain.com:51820
```

If i spin this up, wireguard will create the correct routes on the host system.
It will create two routes:
![image](https://github.com/ngoduykhanh/wireguard-ui/assets/15030658/9753a5b6-3817-46fe-bdef-ab99728faeb0)

10.0.2.0/24 (my friends wg network) is routed through the wg0 interface, as well as `192.168.0.0/24` (my friends home network).
Wireguard will route IPs in those ranges through the tunnel between my friend and me then.
Theres no problem here. The exact same thing happens on the other side.

Of course i assume that a connection was established by my friend before and is kept alive. I could also add an "Endpoint" entry in my config.

So i really don't see why i would have to create a new profile or anything with an IP in my servers subnet. The above configs are all i need, both peers are in completely different subnets, and they can connect to each other just fine and forward traffic without any additional config necessary.

Maybe i am missing your point - then sorry for that - but i still think that the IP Allocation limitation in wireguard-ui (client has to be in the subnet of the server) is a limitation that would not be necessary.

**Comment by relaxibus** [comment]
Just as a reminder: is site to site support now included?

**Comment by HarburGER** [comment]
I would also like to have site to site support ✌🏼

**Comment by ClusterDuckster** [comment]
Hi, I am also running into this.

I think this issue is still open and relevant!

From what I understand @Tarow is completely in the right here and I personally cannot follow the reasoning from @jameswillhoite .

Lets say we have 

- `SITE_A` with a wireguard-ui 
- `SITE_B` that is a router 
    - with a LAN `192.168.0.0/24` 

I want the LAN of `SITE_B` be made available on `SITE_A`. 
The only option is to have the `192.168.0.0/24` in the AllowedIPs line of the Peer block for `SITE_B` of the `wg0.conf` file in `SITE_A`.
You CANNOT add `192.168.0.0/24` to the server interface addresses, as this will allocate this subnet locally on `SITE_A`
You CANNOT only add a static route for `192.168.0.0/24` on `SITE_A` to `wg0`, as wireguard will not know to what peer it should send this packets.
Only the AllowedIP setting will create the static route AND make wireguard aware that this subnet belongs to this peer.

I also want to point out that the wording "IP Allocation" is only used by **wireguard-ui**. The **wireguard** software itself does not use this wording.

I can also offer to make a pull request for that.
In case the safety-guard of "IP Allocation" is still somewhat wanted, I could imagine a popover/modal with a warning could be a compromise?

Basically changing the error "IP 192.168.0.0 does not belong to any network addresses of WireGuard server"
to a Modal: 
> IP 192.168.0.0 does not belong to any network addresses of WireGuard server.
> Are you sure the IP or Subnet is available on the peer?
>
> Yes, I'm sure.    /    Cancel

I would be very happy if I could get some feedback on that.

**Comment by jameswillhoite** [comment]
@ClusterDuckster yes @Tarow is correct above. If you have Wireguard UI set up on both systems, and both generate a config file then that set up will work as long as 
1) each node is in a different subnet (Site A has a LAN subnet of 192.168.0.0/24 and Wireguard Subnet of 10.10.0.0/24 AND Site B has a LAN subnet of 192.168.1.0/24 and Wireguard Subnet of 10.11.0.0/24)
2) If these "Sites" are just machines in a Network and not at the Edge, then a Static Route will need to be added to the Routers to route that subnet to the correct machine

Asking for Site to Site like this in a Single UI is not really Possible because there are many factors outside the UI. It would require someone Technical to add the correct routes. Even my Firewalla doesn't allow 3rd party site to site setup via their UI because they can't control both ends. There is a manual config adjustment required for this to happen. 

Honestly, you can just set up ONE connection and have it Site to Site. As long as you have the Routes and Allowed IPs set up correctly. I have this one connection set up with Site to Site with most of my Family Members homes. They all have routes to me, and I can even forward some to another VPN client depending on rules.

**Comment by Tarow** [comment]
@ClusterDuckster It's been a year but back then i also forked the repo, made some changes and tested it with some machines. And it worked perfectly fine for my use-case.

Just realized i never even pushed the changes to my fork, and didn't think about creating a PR. So my fork is a bit behind the upstream.now, but heres what i did a while ago and in my tests it worked just fine for me:
https://github.com/Tarow/wireguard-ui/commit/3c3e2e951170eaee01d2ca4dfb6eb942e0533dc2

**Comment by ClusterDuckster** [comment]
Wow, thank you both for your quick responses!

So the question if we CAN use wireguard-ui for site-to-site can be answered with yes (with changes by @Tarow).
But I feel like @jameswillhoite raised the question if you SHOULD, from what I read between the lines.

Personally I feel like you should be able to do it by UI, since a lot of end-user routers already offer wireguard site-to-site functionality per UI.

But I also like james suggestion to have the sites in a seperate configuration. I just tried it out, so I now have in `/etc/wireguard`:
- `wg0.conf` -> managed by wireguard-ui 
- `wg_sites.conf` -> manually managed with all site-to-site configs

so running 2 `wg-quick` services on 2 different ports, but I guess I should be able to create routes between them. 
Also an unexpected bonus: the `wg_sites` connections can also be seen in the status page of wireguard-ui.

I will keep trying with that setup and hope I will get my vpn finally to work completely.
Thanks again to the both of you!
